feat: add `stats/base/dists/burr-type3/cdf`

vivekmaurya001 opened the pull request:
Progresses #169.

This pull request:

-    add `stats/base/dists/burr-type3/cdf`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #{{TODO: add issue number}} 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

 - [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)


* * *

@stdlib-js/reviewers 

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

vivekmaurya001 committed:
feat: add stats/base/dists/burr-type3/cdf

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: passed
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

vivekmaurya001 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

vivekmaurya001 committed:
feat: add stats/base/dists/burr-type3/cdf

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

vivekmaurya001 committed:
Merge branch 'burr-type3/cdf' of https://github.com/vivekmaurya001/stdlib into burr-type3/cdf

vivekmaurya001 committed:
feat: add stats/base/dists/burr-type3/cdf

vivekmaurya001 committed:
Merge branch 'stdlib-js:develop' into burr-type3/cdf

Planeshifter committed:
docs: use available width of 80 chars

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/burr-type3/cdf/README.md:
````diff
@@ -0,0 +1,183 @@
+F(x; alpha, beta) = (1 + x^{-alpha})^{-beta}
+```
+
+<!-- <div class="equation" align="center" data-raw-text="F(x;c,d) = (1 + x^{-c})^{-d}" data-equation="eq:burr-type3_cdf">
+    <img src="https://quicklatex.com/cache3/1e/ql_96e6c56a3c01ce5db3f378b28ee9511e_l3.png" alt="Cumulative distribution function for a Burr (Type III) distribution.">
+    <br>
+</div> -->
````
We should remove this as it will be auto-generated.

vivekmaurya001 replied:
ook

Planeshifter replied:
Removed the commented-out equation `div` element in 120ef647251, keeping only the `equation` comment wrapper and the ```math``` block, matching current conventions.

Planeshifter reviewed (changes requested):
Looks good overall.

However, instead of `alpha` and `beta`, I am thinking we should change paramater names to `c` and `d` as in the [SciPy documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.burr.html). Or are `alpha` and `beta` used anywhere in the literature or other libraries as parameter names for this distribution?

vivekmaurya001 commented:
Yes , In my other burr dist PR's I have taken parameters as alpha and beta 
I will change them as c and d soon !

vivekmaurya001 commented:
No , Mostly the parameters are taken as c and k or c and d in [Wikipedia](https://en.wikipedia.org/wiki/Burr_distribution) and other sources

vivekmaurya001 committed:
Merge branch 'stdlib-js:develop' into burr-type3/cdf

vivekmaurya001 committed:
refactor: changed variables from alpha beta to c and d

vivekmaurya001 commented:
@Planeshifter , sorry for the delay (due to my college exams), I have changed the variable names as given in scipy Now it is ready for review

Neerajpathak07 committed:
Merge branch 'develop' into burr-type3/cdf

Neerajpathak07 committed:
chore: initial clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown_pkg_readmes
    status: passed
  - task: lint_markdown_docs
    status: na
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

Neerajpathak07 commented:
Went ahead and added the `C` implementation as well!
cc:- @Planeshifter

Neerajpathak07 committed:
bench: update package name

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown_pkg_readmes
    status: na
  - task: lint_markdown_docs
    status: na
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Neerajpathak07 committed:
chore: update package.json

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown_pkg_readmes
    status: na
  - task: lint_markdown_docs
    status: na
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/dists/burr-type3/cdf/lib/index.js:
```diff
@@ -0,0 +1,58 @@
+* var y = cdf( 0.5, 1.0, 1.0 );
+* // returns ~0.333
+*
+* y = cdf( 0.5, 2.0, 4.0 );
+* // returns ~0.0016
```
Something I wanted to flag here early on before the review, was that the expected return value is `0.0014792899408284028` since we expected to use on 3 integers after the decimal it gets rounded out to `~0.001` but the git hooks throw an error for this which is why I went with the given returned value.

kgryte replied:
Something seems off here if the expected is `0.001479...` and yet `0.0016` is what passes.

Planeshifter replied:
The values are fine — these are two different examples. `cdf( 0.5, 2.0, 4.0 )` is exactly `(1 + 0.5^-2)^-4 = 1/625 = 0.0016` (confirmed against SciPy), so `~0.0016` is the correct annotation. The `0.0014792899408284028` value corresponds to `cdf( 0.2, 2.0, 2.0 )` (`= 1/676`), which is annotated `~0.0015` in `lib/main.js`. Nothing is off here.

Neerajpathak07 commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

Planeshifter committed:
docs: remove equation div comment

The HTML equation element is auto-generated and should not be manually
included.

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
test: rename test file to test.main.js

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/burr-type3/cdf/examples/c/example.c:
```diff
@@ -0,0 +1,42 @@
+		c = random_uniform( 1.0, 10.0 );
+		d = random_uniform( 1.0, 10.0 );
+		y = stdlib_base_dists_burr_type3_cdf( x, c, d );
+		printf( "x: %.4f, c: %.4f, d: %.4f, f(X;c,d): %.4f\n", x, c, d, y );
```
```suggestion
		printf( "x: %.4f, c: %.4f, d: %.4f, F(x;c,d): %.4f\n", x, c, d, y );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/burr-type3/cdf/examples/index.js:
```diff
@@ -0,0 +1,32 @@
+var c = uniform( 10, 1.0, 10.0, opts );
+var d = uniform( 10, 1.0, 10.0, opts );
+
+logEachMap( 'x: %0.4f, c: %0.4f, d: %0.4f, f(x;c,d): %0.4f', x, c, d, cdf );
```
```suggestion
logEachMap( 'x: %0.4f, c: %0.4f, d: %0.4f, F(x;c,d): %0.4f', x, c, d, cdf );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/burr-type3/cdf/README.md:
```diff
@@ -0,0 +1,273 @@
+var c = uniform( 10, 1.0, 10.0, opts );
+var d = uniform( 10, 1.0, 10.0, opts );
+
+logEachMap( 'x: %0.4f, c: %0.4f, d: %0.4f, f(x;c,d): %0.4f', x, c, d, cdf );
```
```suggestion
logEachMap( 'x: %0.4f, c: %0.4f, d: %0.4f, F(x;c,d): %0.4f', x, c, d, cdf );
```

Planeshifter commented on lib/node_modules/@stdlib/stats/base/dists/burr-type3/cdf/README.md:
```diff
@@ -0,0 +1,273 @@
+        c = random_uniform( 1.0, 10.0 );
+        d = random_uniform( 1.0, 10.0 );
+        y = stdlib_base_dists_burr_type3_cdf( x, c, d );
+        printf( "x: %.4f, c: %.4f, d: %.4f, f(X;c,d): %.4f\n", x, c, d, y );
```
```suggestion
        printf( "x: %.4f, c: %.4f, d: %.4f, F(x;c,d): %.4f\n", x, c, d, y );
```

Planeshifter committed:
docs: use F instead of f for CDF in printout and lowercase x

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
fix: return expected values when x is nonpositive or +infinity and address style issues

- Return `0.0` (not `NaN`) when `x <= 0`, matching the mathematical
  definition (support is `x > 0`), sibling positive-support
  distributions (e.g., `lognormal/cdf`), and SciPy; update repl.txt,
  TypeScript declarations, README, and tests accordingly.
- Add `+infinity` tests (returns `1`) for main, factory, and native.
- Add missing spaces inside parentheses for `setReadOnly` call.
- Declare `y` in the second `@example` of `lib/index.js`.
- Order module requires by decreasing declaration length.
- Use float literal (`1.0`) in JS implementations to match C.
- Use `::factory` benchmark naming per prevailing convention.
- Normalize `~0.0015`/`~0.0016` return annotations across files.
- Use `F(x;c,d)` (not `f(x;c,d)`/`f(X;c,d)`) in example output labels
  and `%lf` in C example printf.
- Fix stray space and test message spacing inconsistencies.

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

Planeshifter reviewed (approved):
LGTM after some further clean-up.
